plugins/single_plugins/idle.cpp

```diff
@@ -4,69 +4,297 @@ extern "C"
 }
 
 #include "singleton-plugin.hpp"
+#include "render-manager.hpp"
 #include "output.hpp"
 #include "core.hpp"
 #include "config.hpp"
 #include "output-layout.hpp"
+#include "../cube/cube-control-signal.hpp"
+
+#include <cmath>
+#include <animation.hpp>
+
+#define ZOOM_BASE 1.0
+
+enum screensaver_state
+{
+    SCREENSAVER_DISABLED,
+    SCREENSAVER_RUNNING,
+    SCREENSAVER_STOPPING
+};
 
 class wayfire_idle
 {
+    wf::effect_hook_t render_hook;
+    double rotation = 0.0;
+    wf_duration duration;
+    wf_transition rot_transition, zoom_transition;
     bool idle_enabled = true;
-    wlr_idle_timeout *timeout = NULL;
-    wf::wl_listener_wrapper on_idle, on_resume;
+    screensaver_state state = SCREENSAVER_DISABLED;
+    std::map<wf::output_t*, bool> hook_set;
+    wlr_idle_timeout *timeout_screensaver = NULL;
+    wlr_idle_timeout *timeout_dpms = NULL;
+    wf::wl_listener_wrapper on_idle_screensaver, on_resume_screensaver;
+    wf::wl_listener_wrapper on_idle_dpms, on_resume_dpms;
 
-    wf_option dpms_timeout;
+    wf_option dpms_timeout, screensaver_timeout;
+    wf_option cube_zoom_speed, cube_rotate_speed, cube_zoom_end;
     wf_option_callback dpms_timeout_updated = [=] () {
-        create_timeout(dpms_timeout->as_int());
+        create_dpms_timeout(dpms_timeout->as_int());
+    };
+    wf_option_callback screensaver_timeout_updated = [=] () {
+        create_screensaver_timeout(screensaver_timeout->as_int());
     };
 
     public:
     wayfire_idle()
     {
         dpms_timeout = wf::get_core().config->get_section("idle")
             ->get_option("dpms_timeout", "-1");
+        screensaver_timeout = wf::get_core().config->get_section("idle")
+            ->get_option("screensaver_timeout", "-1");
+        cube_zoom_speed = wf::get_core().config->get_section("idle")
+            ->get_option("cube_zoom_speed", "1000");
+        cube_rotate_speed = wf::get_core().config->get_section("idle")
+            ->get_option("cube_rotate_speed", "0.001");
+        cube_zoom_end = wf::get_core().config->get_section("idle")
+            ->get_option("cube_max_zoom", "1.2");
+
+        render_hook = [=] () { screensaver_frame(); };
 
         dpms_timeout->add_updated_handler(&dpms_timeout_updated);
         dpms_timeout_updated();
+        screensaver_timeout->add_updated_handler(&screensaver_timeout_updated);
+        screensaver_timeout_updated();
+        duration = wf_duration(cube_zoom_speed);
+        
+        for (auto& output : wf::get_core().output_layout->get_outputs())
```

Comment:
no need for this. Operator [] will return the default value for bool (i.e false) if the element is not in the map.

Reply:
Ok.